### PR TITLE
vhost_rs: some changes according to vhost user spec 

### PR DESCRIPTION
**Pull request by yangzhon**
As @sboeuf said in #173, i split out vhost user changes to this PR, the older PR will be as pure vhost-user-blok changes.

There are some bugs in current vhost_rs code, and read_config()/write_config() are not compatible with vhost user config info spec. I changed those according to vhost user spec, please help review those patches, thanks. 

@sameo @sboeuf @rbradford @chao-p

### CODE REVIEW

**Comment by rbradford** [comment]
@yangzhon You need to run rustfmt and squash into the appropriate commits.

**Comment by yangzhon** [comment]
@rbradford , 

Thanks for your reminder, i have squashed the commit with "cargo fmt" tool check, thanks.

**Comment by rbradford** [comment]
Tapping @sboeuf  to approve.

File: vhost_rs/src/vhost_user/message.rs [hunk 1]
```diff
@@ -575,10 +575,9 @@ impl VhostUserMsgValidator for VhostUserConfig {
     fn is_valid(&self) -> bool {
         if (self.flags & !VhostUserConfigFlags::all().bits()) != 0 {
             return false;
-        } else if self.offset < VHOST_USER_CONFIG_OFFSET
-            || self.offset >= VHOST_USER_CONFIG_SIZE
+        } else if self.offset >= VHOST_USER_CONFIG_SIZE
```
Comment (sboeuf) [inline vhost_rs/src/vhost_user/message.rs]: From your commit message, you explain this is specific to `virtio-mmio`. Could you detail what is different in this case between `virtio-mmio` and `virtio-pci`, and does it break `virtio-pci`?

Second thing, this crate is going to be used by VMM that supports `virtio-mmio` only, hence we need it to work for both transport layers. Rust-vmm will never accept this crate if it's exclusively `virtio-pci` oriented.

File: vhost_rs/src/vhost_user/master.rs [hunk 2]
```diff
@@ -603,7 +603,7 @@ impl MasterInternal {
     #[inline]
     fn new_request_header(request: MasterReq, size: u32) -> VhostUserMsgHeader<MasterReq> {
         // TODO: handle NEED_REPLY flag
-        VhostUserMsgHeader::new(request, 0, size)
+        VhostUserMsgHeader::new(request, 0x1, size)
```
Comment (sboeuf) [inline vhost_rs/src/vhost_user/master.rs]: Please change the commit message here, as the point is not about what QEMU is doing, but really what we actually support.

Comment (yangzhon) [reply vhost_rs/src/vhost_user/master.rs]: Yes, thanks, i will change it.

File: vhost_rs/src/vhost_user/message.rs (hunk 1 above)
Comment (yangzhon) [reply vhost_rs/src/vhost_user/message.rs]: @sboeuf ,  as for offset definition in virtio-mmio, please ref **4.2.2 MMIO Device Register Layout** in virtio spec 1.1

I also list the key info below:
virtio-mmio:
Configuration space
Device-specific configuration space starts at the **offset 0x100** and is accessed with byte alignment. Its meaning and size depend on the device and the driver

By the way, i talked with Liujiang, and we agreed this changes and will add offset in virtio-mmio side, not vhost user protocol.

Liujiang will release new vhost_rs release and one guy from redhat has submitted one patch for this, please ref https://github.com/jiangliu/vhost/commit/b5db8ec07d4914cca708a64af96e361a66f8a47a

**Comment by yangzhon** [comment]
@sboeuf @sameo @rbradford 

Liujiang will release new vhost_rs release, but this need util sys and vm-memory updatation. 
https://github.com/jiangliu/vhost/commits/v1

My vhost_rs changes has 5 patches, and his new release has covered 3 patches. The get_config() changes in his release only covered slave side, not master side, so my get_config() patch does not conflict with him.

Do we have plan to upgrade this version ? if not, i will continue to submit my patch here, thanks.

**Comment by sameo** [comment]
@yangzhon 
> Do we have plan to upgrade this version ? if not, i will continue to submit my patch here, thanks.

We should keep in sync with the Gerry's vhost_rs version, because we can assume this is what's going to be merged into rust-vmm. Are you submitting your changes to Gerry's PR as well?

**Comment by yangzhon** [comment]
> We should keep in sync with the Gerry's vhost_rs version, because we can assume this is what's going to be merged into rust-vmm. Are you submitting your changes to Gerry's PR as well?

@sameo ,

Since the v1 has been released from Liujiang, i will submit one dedicated PR to vhost_rs github for the left 2 patches. I talked with Liujiang today, and he hope i will add test case and help him to verify get_config)/set_config() functions. 

If we upgrade vhost_rs to v1, we also need to update util system and vm-memory crates. In fact, the new vhost_rs release did not be fully verified by them. 

Maybe we can merge those patches into cloud-hypervisor firstly, and then i will send another PR to vhost_rs.  Once this PR is merged into cloud-hypervisor, i will submit vhost-user-blk PR. thanks.

**Comment by yangzhon** [comment]
@sboeuf ,

Your patch, c6feb03dc08dfc6f4faea211811361228e8f86f8, which will make my get_config() not work in vhost-user-blk. 

@sameo ,

I also submitted one PR to vhost github, Liujiang will deal with this PR this tonight.
https://github.com/jiangliu/vhost/pull/3

thanks

**Comment by yangzhon** [comment]
> @sboeuf ,
> 
> Your patch, [c6feb03](https://github.com/intel/cloud-hypervisor/commit/c6feb03dc08dfc6f4faea211811361228e8f86f8), which will make my get_config() not work in vhost-user-blk.
> 
> @sameo ,
> 
> I also submitted one PR to vhost github, Liujiang will deal with this PR this tonight.
> [jiangliu/vhost#3](https://github.com/jiangliu/vhost/pull/3)
> 
> thanks

@sboeuf 

I will address comments from Liujiang for https://github.com/jiangliu/vhost/pull/3,  do you have plan to push c6feb03 patch into vhost github? thanks

File: vhost_rs/src/vhost_user/message.rs (hunk 1 above)
Comment (sboeuf) [reply vhost_rs/src/vhost_user/message.rs]: Yes that's what I was expecting. The change has to happen on the transport layer implementation, it should be independent from vhost-user.

**Comment by sboeuf** [comment]
@yangzhon 
>Your patch, c6feb03, which will make my get_config() not work in vhost-user-blk.

Could you be more specific about why it's not working? Have you identified the root cause. It might be a bug in the `vhost-user-blk backend` that you're using.

>I will address comments from Liujiang for jiangliu/vhost#3, do you have plan to push c6feb03 patch into vhost github? thanks

I don't, but you can go ahead and push it to Gerry's branch, so that later we don't have gaps between Gerry's version and the cloud-hypervisor's one.

**Pull request merged**